### PR TITLE
Add support for non-hypercube grids

### PR DESCRIPTION
**Pull request by tobiasleibner**
This adds support for arbitrary rectangular grids (that are parallel to the coordinate axes) to PeriodicGridView and adds tests. 
Also contains a commit that adds ApplyOn::NonPeriodicBoundaryIntersections (tests intersection.boundary() && !intersection.neighbor()).

### CODE REVIEW

**Comment by tobiasleibner** [comment]
Oh, I checked again and actually there are some more commits contained:
A change to functions.expression to create() with non-empty gradients and the removal of the RungeKuttaTimeStepper. I can split it in several pull requests if you want.

File: dune/stuff/functions/expression.hh [hunk 1]
```diff
@@ -303,6 +340,21 @@ private:
       }
   } // ... build_gradients(...)
 
+  template< int rows, int cols >
+  void build_gradients(const std::string variable,
+                       Dune::FieldMatrix< std::string, rows, cols > jacobian)
+  {
+    assert(jacobian.rows == 0 || jacobian.rows >= dimRange);
```
Comment (ftalbrecht) [inline dune/stuff/functions/expression.hh]: This is not how a `FieldMatrix` works. It has a static size of `rows` times `cols`, so no checks needed.

File: dune/stuff/functions/expression.hh [hunk 2]
```diff
@@ -242,6 +250,35 @@ public:
...
+    build_gradients(variable, jacobian);
+  }
+
+  template< int rows, int cols >
+  Expression(const std::string variable,
+             const std::vector< std::string > expressions,
+             const size_t ord = default_config().get< size_t >("order"),
+             const std::string nm = static_id(),
+             const Dune::FieldMatrix< std::string, rows, cols > jacobian
+                = Dune::FieldMatrix< std::string, 0, 0 >())
```
Comment (ftalbrecht) [inline dune/stuff/functions/expression.hh]: This should not compile, as both matrices have a different type.

**Comment by ftalbrecht** [comment]
I think you should separate the two things, as I am not sure the expression gradient is correct yet (see my comments). The extension of the periodic grid view looks good, though, and could be merged imho.

File: dune/stuff/functions/expression.hh (hunk 2 above)
Comment (tobiasleibner) [inline dune/stuff/functions/expression.hh]: I think it did compile... I will check again. Shouldn't the compiler be able to figure out that rows = cols = 0 if the default value is used?

File: dune/stuff/functions/expression.hh (hunk 1 above)
Comment (tobiasleibner) [inline dune/stuff/functions/expression.hh]: Actually, the check is needed because the method is templated by rows and cols, I think.

**Comment by tobiasleibner** [comment]
I reverted the changes to functions.expression and opened a separate pull request #49.

**Pull request merged**